### ISSUE TITLE
sdefl not compatible with firefox built-in deflate method

### ISSUE THREAD

**Issue by grassroot72** [vurtun/lib#34]
I tried to use sdefl as the built-in compression tool in my webserver to deflate the http messages . It seems that the compressed data is not compatible with the firefox built-in decompression method. The firefox continues to issue complaints of "The page you are trying to view cannot be shown because it uses an invalid or unsupported form of compression." I have tried a webkit based browser too, it keeps showing junk all the time.

I had used the earlier version of sdefl when it was only ~300 lines of code (before you moved the sdefl code to lib), it worked perfectly with my webserver. The sdefl is a great library, the reason for me to move to your new sdefl is to adapt to get higher compression ratio.

Have you made this incompatible change on purpose? If not, would you make it compatible with firefox built-in deflate method?

**Comment by vurtun** [comment]
Hey, incompatibility was definitely not on purpose. What API are you using? `sdeflate` or `zsdeflate`? I tested `sdeflate` by decompressing using `libdeflate` and as far as I can tell everything worked correctly. A possibility is that while I pulled out the zlib header into `zsdeflate` from `sdefl_compr` introduced a bug. Also which deflate version did you use previously the static or the adaptive huffman variant?

**Comment by grassroot72** [comment]
I tried both sdeflate and zsdeflate of the new sdefl library, they both give incompatible compressed data. The old sdefl I had used was the static version which gave compatible compressed data.

**Comment by vurtun** [comment]
Have you tried other alternatives like `miniz` or `libdeflate`.  Would be interesting to know if the problem is specific to `sdefl` or if firefox only supports static huffman compression.

**Comment by grassroot72** [comment]
I will try those libraries and update you later, it is quite late in the night my time :-)

**Comment by vurtun** [comment]
Ok, I think I found the problem. In my precode calculation I forgot to only send the actual number of used literals and offsets and instead just send all of them. While this works for `libdeflate`, `sinfl` and `miniz` is not really conformant. I will fix this issue as soon as possible. For now I would recommend using the static huffman table for now until I fix this version (hopefully tomorrow).

**Comment by grassroot72** [comment]
I tried libdeflate by replacing the inflate part of defl_test.c with something like this
`result_code = libdeflate_deflate_decompress(d, comp, len, decomp, size, &out_size);`
, it yielded correct uncompressed data content and correct uncompressed data length.

Yes, you are probably right. I also doubt if libdeflate is conformant, because when I looked into RFC7230, the chapter `4.2.2.  Deflate Coding` says 

> The "deflate" coding is a "zlib" data format [RFC1950] containing a
   "deflate" compressed data stream [RFC1951] that uses a combination of
   the Lempel-Ziv (LZ77) compression algorithm and Huffman coding.

That means firefox should support Huffman coding with its deflate encoding method.
Thanks again for your great work! 
Hope to hear the good news from you soon :-)

**Comment by vurtun** [comment]
Hey, I created a test version here: https://gist.github.com/vurtun/69d86630528bbfb264101b0f1b936d8d for `sdefl.h`. Could you try this out and see if it works. I don't want to push this yet since if the problem lies somewhere else I don't want to add the additional complexity.

**Comment by grassroot72** [comment]
The test version works perfectly with my webserver. Great!!!!!

I have also done some benchmark test with the 'incompatiable(non-conformant)' version and 'compatible(conformant) version.

**Uncompressed file name: bench.html**
**Uncompressed file size: 1577**

With **sdefl.h (incompatible version):**
> lvl: 0 size: 853, compr: 2.20ms (0.72MB/s) decompr: 0.51ms (3.10MB/s) ratio: 54.090044% 
lvl: 1 size: 848, compr: 1.54ms (1.02MB/s) decompr: 0.53ms (3.00MB/s) ratio: 53.772987% 
lvl: 2 size: 841, compr: 1.57ms (1.00MB/s) decompr: 0.47ms (3.39MB/s) ratio: 53.329106% 
lvl: 3 size: 841, compr: 1.64ms (0.96MB/s) decompr: 0.47ms (3.37MB/s) ratio: 53.329106% 
lvl: 4 size: 841, compr: 1.61ms (0.98MB/s) decompr: 0.49ms (3.22MB/s) ratio: 53.329106% 
lvl: 5 size: 838, compr: 0.39ms (4.04MB/s) decompr: 0.11ms (14.76MB/s) ratio: 53.138871% 
lvl: 6 size: 838, compr: 0.39ms (4.08MB/s) decompr: 0.11ms (14.81MB/s) ratio: 53.138871% 
lvl: 7 size: 838, compr: 0.39ms (4.08MB/s) decompr: 0.11ms (14.82MB/s) ratio: 53.138871%

With **test_sdefl.h (compatible version):**
> lvl: 0 size: 848, compr: 2.12ms (0.74MB/s) decompr: 0.50ms (3.14MB/s) ratio: 53.772987% 
lvl: 1 size: 843, compr: 1.03ms (1.52MB/s) decompr: 0.11ms (14.61MB/s) ratio: 53.455929% 
lvl: 2 size: 836, compr: 0.36ms (4.34MB/s) decompr: 0.11ms (14.86MB/s) ratio: 53.012048% 
lvl: 3 size: 836, compr: 0.36ms (4.35MB/s) decompr: 0.11ms (14.91MB/s) ratio: 53.012048% 
lvl: 4 size: 836, compr: 0.36ms (4.34MB/s) decompr: 0.11ms (14.69MB/s) ratio: 53.012048% 
lvl: 5 size: 833, compr: 0.39ms (4.02MB/s) decompr: 0.11ms (14.85MB/s) ratio: 52.821814% 
lvl: 6 size: 833, compr: 0.39ms (4.02MB/s) decompr: 0.13ms (11.79MB/s) ratio: 52.821814% 
lvl: 7 size: 833, compr: 0.39ms (4.05MB/s) decompr: 0.11ms (14.94MB/s) ratio: 52.821814%

The changes you have mode not only solved the problem, but also improved the compression rate about 0.5%!!!
Nice :-)
I haven't done a thorough benchmark for this, but it seems level 4 can be chosen to be the default compression level. It is only my rough observation. You are the authority, it needs your discretion.

I see that you also have a webserver, quite efficient. If not a trouble for you, I would like to invite you to help me improve my webserver.
[Maestro](https://github.com/grassroot72/Maestro) (Need to renew README.md ...)

Have a good day!

Cheers,
Edward

**Comment by vurtun** [comment]
Yay success. Sounds good. I merged the test version. In `sdefl.h` I defined `SDEFL_LVL_DEF` as `5` as default compression level since at that point a second match find iteration is called to look for better compression. How good that is depends somewhat on the data being compressed. If for you use case there is no real advantage for an additional iteration `4` is probably a bit of a performance boost.

Yeah I have a minimal web server single header library as well. However it focuses is more in embedding into an application and send out for example performance data via web sockets into a web browser and less on of actual multi-client serving. I can take a quick look over Maestro but I can't promise anything.
